--- a/target/linux/ramips/image/mt7620.mk
+++ b/target/linux/ramips/image/mt7620.mk
@@ -784,11 +784,11 @@
 
 define Device/phicomm_psg1218a
   SOC := mt7620a
-  IMAGE_SIZE := 7872k
+  IMAGE_SIZE := 16064k
   DEVICE_VENDOR := Phicomm
   DEVICE_MODEL := PSG1218
   DEVICE_VARIANT:= Ax
   DEVICE_PACKAGES := kmod-mt76x2
-  SUPPORTED_DEVICES += psg1218 psg1218a
+  SUPPORTED_DEVICES += psg1218a
 endef
 TARGET_DEVICES += phicomm_psg1218a
